packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -56,16 +57,23 @@ function UnconnectedNavigatorProvider(
 	props: WordPressComponentProps< NavigatorProviderProps, 'div' >,
 	forwardedRef: ForwardedRef< any >
 ) {
-	const { initialPath, children, className, ...otherProps } =
-		useContextSystem( props, 'NavigatorProvider' );
+	const {
+		initialPath,
+		location: locationProp,
+		onChange,
+		children,
+		className,
+		...otherProps
+	} = useContextSystem( props, 'NavigatorProvider' );
+	const [ location, updateLocation ] = useControlledValue( {
```

Comment:
Does it make sense to keep both `location` and `locationHistory` internal states? I'm afraid it would cause extra re-renders and potentially some race conditions too. Maybe we can just keep the location history as a ref, and update `navigatorContextValue` accordingly? (I know that this would probably also require refactoring the logic around the new `useEffect`)